I'd like to be able to specify e.g. `--property-variables` in / per module, but GHC makes it not nice:

```
src/Numeric/Interval/NonEmpty/Internal.hs:8:1: warning: [-Wunrecognised-pragmas]
    Unrecognised pragma
  |
8 | {-# OPTIONS_DOCSPEC --property-variables="foo bar" #-}
  | ^^^
```

--- a/cabal-docspec/MANUAL.md
+++ b/cabal-docspec/MANUAL.md
@@ -83,6 +83,10 @@ However, in this list we mostly only list and show the --option version of them.
 :   Message to return when the evaluation is timed out.
     Default is **\* Hangs forever \***.
 
+**\--ghci-rtsopts** *options*
+
+:   RTS options for GHCi process
+
 **\--skip-properties**
 
 :   Skip properties.
@@ -178,6 +182,8 @@ There are some examples using explanatory comments,
 Some examples are illustrating non-termination,
 therefore short **\--timeout** is justified.
 Yet, it have to be long enough so terminating examples have time to run.
+We also set RTS options, reducing the maximum stack to
+make stack overflow exceptions occur earlier.
 Few examples are using symbols from *mtl*, *deepseq* and *bytestring* packages,
 we make them available.
 Finally, some modules are documented with no-Prelude assumption,
@@ -188,7 +194,8 @@ therefore we have to turn it off.
         -I $PWD/includes \
         --no-cabal-plan \
         --strip-comments \
-        --timeout 2.5 \
+        --timeout 2 \
+        --ghci-rtsopts "-K500K" \
         --extra-package=mtl --extra-package=deepseq --extra-package=bytestring \
         -XNoImplicitPrelude \
         libraries/base/base.cabal

--- a/cabal-docspec/MANUAL.md
+++ b/cabal-docspec/MANUAL.md
@@ -638,7 +638,7 @@ For example even
 is enough.
 However, without libraries actually being built, cabal-docspec won't work.
 
-Q: Does Doctest's --fast have an equivalent in cabal-docspec?
+Q: Does Doctest's \--fast have an equivalent in cabal-docspec?
 -------------------------------------------------------------
 
 No, cabal-doctest doesn't need one.

--- a/cabal-docspec/MANUAL.md
+++ b/cabal-docspec/MANUAL.md
@@ -83,10 +83,27 @@ However, in this list we mostly only list and show the --option version of them.
 :   Message to return when the evaluation is timed out.
     Default is **\* Hangs forever \***.
 
+**\--skip-properties**
+
+:   Skip properties.
+
+**\--simple-properties**
+
+:   Evaluate **prop> expr x y** using **quickCheck (expr x y)**.
+    Requires *QuickCheck* package in the plan.
+
+**\--property-variables** *varlist*
+
+:   Variables to automatically quantify over in properties.
+
 **\-X** *extension*
 
 :   Language extension to start GHCi session with.
     Can be specified multiple times.
+
+**\-I** *directory*
+
+:   Add *directory* to the directory search list for **#include** files.
 
 **\--phase1**
 
@@ -139,6 +156,10 @@ through fields in a .cabal file.
 
 :    A (space separated) list of extra packages. See **\--extra-package**.
 
+**x-docspec-property-variables:** *[VAR]*...
+
+:    A (space separated) list of property variables. See **\--property-variables**.
+
 EXAMPLES
 ========
 
@@ -156,16 +177,19 @@ There are some examples using explanatory comments,
 **\--strip-comments** makes them work.
 Some examples are illustrating non-termination,
 therefore short **\--timeout** is justified.
-Few examples are using symbols from *mtl* and *deepseq* packages,
+Yet, it have to be long enough so terminating examples have time to run.
+Few examples are using symbols from *mtl*, *deepseq* and *bytestring* packages,
 we make them available.
 Finally, some modules are documented with no-Prelude assumption,
 therefore we have to turn it off.
 
-    cabal-docspec -w $PWD/_build/stage1/bin/ghc \
+    cabal-docspec \
+        -w $PWD/_build/stage1/bin/ghc \
+        -I $PWD/includes \
         --no-cabal-plan \
         --strip-comments \
-        --timeout 2 \
-        --extra-package=mtl --extra-package=deepseq \
+        --timeout 2.5 \
+        --extra-package=mtl --extra-package=deepseq --extra-package=bytestring \
         -XNoImplicitPrelude \
         libraries/base/base.cabal
 
@@ -392,10 +416,15 @@ If a line contains three dots and additional content, the three dots will match 
 QuickCheck properties
 ---------------------
 
-**NOTE:** cabal-docspec doesn't check properties at the moment. Details may change.
-
 Haddock (since version 2.13.0) has markup support for properties
-Doctest can verify properties with QuickCheck.
+cabal-docspec can verify properties with QuickCheck.
+Note: there are some differences with Doctest.
+
+By default properties are skipped. This is a **\--skip-properties** behaviour.
+cabal-docspec has a simple mechanism to evaluate properties
+enabled by **\--simple-properties**.
+For it to work *QuickCheck* package have to be in the install plan.
+
 A simple property looks like this:
 
 ```haskell
@@ -403,11 +432,43 @@ A simple property looks like this:
 -- prop> \xs -> sort xs == (sort . sort) (xs :: [Int])
 ```
 
-The lambda abstraction is optional and can be omitted:
+The lambda abstraction is not optional by default.
+cabal-docspec will automatically qualify over variables
+passed in with **\--property-variables** command line flag.
+
+With **--property-variables xs** the following will work:
 
 ```haskell
 -- |
 -- prop> sort xs == (sort . sort) (xs :: [Int])
+```
+
+Doctest uses a hack to find which variables are free in the the expression,
+cabal-docspec approach is more deterministic, it doesn't try to infer anything.
+
+Also contrary to *Doctest* cabal-docspec doesn't use **polyQuickCheck** trick.
+Therefore some false properties may pass
+
+```haskell
+quickCheck $ \xs -> reverse xs === xs
++++ OK, passed 100 tests.
+```
+
+That property passes because list element types defaults to **()**.
+To avoid defaulting you may override defaults class resoltuion in **$setup** block
+
+```haskell
+-- $setup
+-- >>> default (Integer, Double)
+```
+
+Then the above property will fail:
+
+```haskell
+quickCheck $ \xs -> reverse xs === xs
+*** Failed! Falsified (after 4 tests and 4 shrinks):    
+[1,0]
+[0,1] /= [1,0]
 ```
 
 A complete example that uses setup code is below:
@@ -430,21 +491,6 @@ fib :: Int -> Int
 fib 0 = 0
 fib 1 = 1
 fib n = fib (n - 1) + fib (n - 2)
-```
-
-If you see an error like the following, ensure that *QuickCheck* is a dependency of the test-suite or executable running docspec (to be corrected).
-
-```haskell
-<interactive>:39:3:
-    Not in scope: ‘polyQuickCheck’
-    In the splice: $(polyQuickCheck (mkName "doctest_prop"))
-
-<interactive>:39:3:
-    GHC stage restriction:
-      ‘polyQuickCheck’ is used in a top-level splice or annotation,
-      and must be imported, not defined locally
-    In the expression: polyQuickCheck (mkName "doctest_prop")
-    In the splice: $(polyQuickCheck (mkName "doctest_prop"))
 ```
 
 Hiding examples from Haddock
@@ -520,6 +566,10 @@ All warnings are enabled by default.
 **-Werror-in-setup**
 
 :   There was an error in evaluting **$setup**.
+
+**-Wskipped-property**
+
+:   Warn about properties when **\--skip-properties** (the default) is enabled.
 
 KNOWN BUGS AND INFECILITIES
 ===========================

--- a/cabal-docspec/Makefile
+++ b/cabal-docspec/Makefile
@@ -1,4 +1,4 @@
-VERSION=0.0.0.20210110
+VERSION=0.0.0.20210111
 
 cabal-docspec.1 : MANUAL.md
 	echo '.TH CABAL-DOCSPEC 1 "January 10, 2021" "cabal-docspec $(VERSION)" "Cabal Extras"' > cabal-docspec.1

--- a/cabal-docspec/cabal-docspec.1
+++ b/cabal-docspec/cabal-docspec.1
@@ -689,7 +689,7 @@ cabal build --dry-run
 is enough.
 However, without libraries actually being built, cabal-docspec won\[cq]t
 work.
-.SS Q: Does Doctest\[cq]s \[en]fast have an equivalent in cabal-docspec?
+.SS Q: Does Doctest\[cq]s --fast have an equivalent in cabal-docspec?
 .PP
 No, cabal-doctest doesn\[cq]t need one.
 The library code is loaded as pre-compiled object code, not interpreted

--- a/cabal-docspec/cabal-docspec.1
+++ b/cabal-docspec/cabal-docspec.1
@@ -88,9 +88,23 @@ On the other hand, too short timeout may cause false negatives.
 Message to return when the evaluation is timed out.
 Default is \f[B]* Hangs forever *\f[R].
 .TP
+\f[B]--skip-properties\f[R]
+Skip properties.
+.TP
+\f[B]--simple-properties\f[R]
+Evaluate \f[B]prop> expr x y\f[R] using \f[B]quickCheck (expr x y)\f[R].
+Requires \f[I]QuickCheck\f[R] package in the plan.
+.TP
+\f[B]--property-variables\f[R] \f[I]varlist\f[R]
+Variables to automatically quantify over in properties.
+.TP
 \f[B]-X\f[R] \f[I]extension\f[R]
 Language extension to start GHCi session with.
 Can be specified multiple times.
+.TP
+\f[B]-I\f[R] \f[I]directory\f[R]
+Add \f[I]directory\f[R] to the directory search list for
+\f[B]#include\f[R] files.
 .TP
 \f[B]--phase1\f[R]
 Stop after the first phase.
@@ -151,18 +165,21 @@ There are some examples using explanatory comments,
 \f[B]--strip-comments\f[R] makes them work.
 Some examples are illustrating non-termination, therefore short
 \f[B]--timeout\f[R] is justified.
-Few examples are using symbols from \f[I]mtl\f[R] and \f[I]deepseq\f[R]
-packages, we make them available.
+Yet, it have to be long enough so terminating examples have time to run.
+Few examples are using symbols from \f[I]mtl\f[R], \f[I]deepseq\f[R] and
+\f[I]bytestring\f[R] packages, we make them available.
 Finally, some modules are documented with no-Prelude assumption,
 therefore we have to turn it off.
 .IP
 .nf
 \f[C]
-cabal-docspec -w $PWD/_build/stage1/bin/ghc \[rs]
+cabal-docspec \[rs]
+    -w $PWD/_build/stage1/bin/ghc \[rs]
+    -I $PWD/includes \[rs]
     --no-cabal-plan \[rs]
     --strip-comments \[rs]
-    --timeout 2 \[rs]
-    --extra-package=mtl --extra-package=deepseq \[rs]
+    --timeout 2.5 \[rs]
+    --extra-package=mtl --extra-package=deepseq --extra-package=bytestring \[rs]
     -XNoImplicitPrelude \[rs]
     libraries/base/base.cabal
 \f[R]
@@ -434,12 +451,16 @@ will match anything \f[I]within that line\f[R]:
 .fi
 .SS QuickCheck properties
 .PP
-\f[B]NOTE:\f[R] cabal-docspec doesn\[cq]t check properties at the
-moment.
-Details may change.
+Haddock (since version 2.13.0) has markup support for properties
+cabal-docspec can verify properties with QuickCheck.
+Note: there are some differences with Doctest.
 .PP
-Haddock (since version 2.13.0) has markup support for properties Doctest
-can verify properties with QuickCheck.
+By default properties are skipped.
+This is a \f[B]--skip-properties\f[R] behaviour.
+cabal-docspec has a simple mechanism to evaluate properties enabled by
+\f[B]--simple-properties\f[R].
+For it work \f[I]QuickCheck\f[R] package have to be in the install plan.
+.PP
 A simple property looks like this:
 .IP
 .nf
@@ -449,12 +470,53 @@ A simple property looks like this:
 \f[R]
 .fi
 .PP
-The lambda abstraction is optional and can be omitted:
+The lambda abstraction is not optional by default.
+cabal-docspec will automatically qualify over variables passed in with
+\f[B]--property-variables\f[R] command line flag.
+.PP
+With \f[B]\[en]property-variables xs\f[R] the following will work.
 .IP
 .nf
 \f[C]
 -- |
 -- prop> sort xs == (sort . sort) (xs :: [Int])
+\f[R]
+.fi
+.PP
+We chose to require explicit list of varaibles in cabal-docspec to avoid
+surprises triggered by implicit behaviour.
+Doctest uses a hack to find which variables are free in the the
+expression, cabal-docspec approach is more deterministic.
+.PP
+Also contrary to \f[I]Doctest\f[R] cabal-docspec doesn\[cq]t use
+\f[B]polyQuickCheck\f[R] trick.
+Therefore some false properties may pass
+.IP
+.nf
+\f[C]
+quickCheck $ \[rs]xs -> reverse xs === xs
++++ OK, passed 100 tests.
+\f[R]
+.fi
+.PP
+To avoid defaulting to (), you may override defaults class resoltuion in
+\f[B]$setup\f[R] block
+.IP
+.nf
+\f[C]
+-- $setup
+-- >>> default (Integer, Double)
+\f[R]
+.fi
+.PP
+Then the above property will fail, as expected:
+.IP
+.nf
+\f[C]
+quickCheck $ \[rs]xs -> reverse xs === xs
+*** Failed! Falsified (after 4 tests and 4 shrinks):    
+[1,0]
+[0,1] /= [1,0]
 \f[R]
 .fi
 .PP
@@ -479,25 +541,6 @@ fib :: Int -> Int
 fib 0 = 0
 fib 1 = 1
 fib n = fib (n - 1) + fib (n - 2)
-\f[R]
-.fi
-.PP
-If you see an error like the following, ensure that \f[I]QuickCheck\f[R]
-is a dependency of the test-suite or executable running docspec (to be
-corrected).
-.IP
-.nf
-\f[C]
-<interactive>:39:3:
-    Not in scope: \[oq]polyQuickCheck\[cq]
-    In the splice: $(polyQuickCheck (mkName \[dq]doctest_prop\[dq]))
-
-<interactive>:39:3:
-    GHC stage restriction:
-      \[oq]polyQuickCheck\[cq] is used in a top-level splice or annotation,
-      and must be imported, not defined locally
-    In the expression: polyQuickCheck (mkName \[dq]doctest_prop\[dq])
-    In the splice: $(polyQuickCheck (mkName \[dq]doctest_prop\[dq]))
 \f[R]
 .fi
 .SS Hiding examples from Haddock
@@ -572,6 +615,10 @@ C preprocessor (\f[I]cpphs\f[R]) warnings.
 .TP
 \f[B]-Werror-in-setup\f[R]
 There was an error in evaluting \f[B]$setup\f[R].
+.TP
+\f[B]-Wskipped-property\f[R]
+Warn about properties when \f[B]--skip-properties\f[R] (the default) is
+enabled.
 .SH KNOWN BUGS AND INFECILITIES
 .PP
 Properties (\f[B]prop>\f[R]) are recognized but not evaluated.

--- a/cabal-docspec/cabal-docspec.1
+++ b/cabal-docspec/cabal-docspec.1
@@ -1,4 +1,4 @@
-.TH CABAL-DOCSPEC 1 "January 10, 2021" "cabal-docspec 0.0.0.20210110" "Cabal Extras"
+.TH CABAL-DOCSPEC 1 "January 10, 2021" "cabal-docspec 0.0.0.20210111" "Cabal Extras"
 .SH NAME
 .PP
 cabal-docspec - another doctest for Haskell
@@ -88,6 +88,9 @@ On the other hand, too short timeout may cause false negatives.
 Message to return when the evaluation is timed out.
 Default is \f[B]* Hangs forever *\f[R].
 .TP
+\f[B]--ghci-rtsopts\f[R] \f[I]options\f[R]
+RTS options for GHCi process
+.TP
 \f[B]--skip-properties\f[R]
 Skip properties.
 .TP
@@ -145,6 +148,10 @@ allow configuration of cabal-docspec per component under test.
 \f[B]x-docspec-extra-packages:\f[R] \f[I][PKG]\f[R]\&...
 A (space separated) list of extra packages.
 See \f[B]--extra-package\f[R].
+.TP
+\f[B]x-docspec-property-variables:\f[R] \f[I][VAR]\f[R]\&...
+A (space separated) list of property variables.
+See \f[B]--property-variables\f[R].
 .SH EXAMPLES
 .PP
 A simplest example, which should work for most packages is to run
@@ -166,6 +173,8 @@ There are some examples using explanatory comments,
 Some examples are illustrating non-termination, therefore short
 \f[B]--timeout\f[R] is justified.
 Yet, it have to be long enough so terminating examples have time to run.
+We also set RTS options, reducing the maximum stack to make stack
+overflow exceptions occur earlier.
 Few examples are using symbols from \f[I]mtl\f[R], \f[I]deepseq\f[R] and
 \f[I]bytestring\f[R] packages, we make them available.
 Finally, some modules are documented with no-Prelude assumption,
@@ -178,7 +187,8 @@ cabal-docspec \[rs]
     -I $PWD/includes \[rs]
     --no-cabal-plan \[rs]
     --strip-comments \[rs]
-    --timeout 2.5 \[rs]
+    --timeout 2 \[rs]
+    --ghci-rtsopts \[dq]-K500K\[dq] \[rs]
     --extra-package=mtl --extra-package=deepseq --extra-package=bytestring \[rs]
     -XNoImplicitPrelude \[rs]
     libraries/base/base.cabal

--- a/cabal-docspec/cabal-docspec.cabal
+++ b/cabal-docspec/cabal-docspec.cabal
@@ -1,6 +1,6 @@
 cabal-version:      2.2
 name:               cabal-docspec
-version:            0.0.0.20210110
+version:            0.0.0.20210111
 synopsis:           Run examples in your docs
 category:           Development
 description:

--- a/cabal-docspec/cabal-docspec.cabal
+++ b/cabal-docspec/cabal-docspec.cabal
@@ -8,13 +8,17 @@ description:
   rely on @ghc@ library.
 
 license:            GPL-2.0-or-later
-license-files:      LICENSE LICENSE.GPLv2 LICENSE.GPLv3
+license-files:
+  LICENSE
+  LICENSE.GPLv2
+  LICENSE.GPLv3
+
 author:             Oleg Grenrus <oleg.grenrus@iki.fi>
 maintainer:         Oleg Grenrus <oleg.grenrus@iki.fi>
 tested-with:        GHC ==8.6.5 || ==8.8.4 || ==8.10.2
 extra-source-files:
-  Changelog.md
   cabal-docspec.1
+  Changelog.md
 
 source-repository head
   type:     git
@@ -25,7 +29,7 @@ library internal-cpphs
   default-language: Haskell2010
   hs-source-dirs:   cpphs
   build-depends:
-    , base         >=4.12   && <5
+    , base         >=4.12    && <5
     , directory    ^>=1.3.0.2
     , polyparse    ^>=1.13
     , time-compat  ^>=1.9.4
@@ -59,6 +63,7 @@ library cabal-docspec-internal
     CabalDocspec.Doctest.Example
     CabalDocspec.Doctest.Extract
     CabalDocspec.Doctest.Parse
+    CabalDocspec.ExprVars
     CabalDocspec.GHCi
     CabalDocspec.Lexer
     CabalDocspec.Located
@@ -118,7 +123,20 @@ executable cabal-docspec
   default-language: Haskell2010
   hs-source-dirs:   cli
   main-is:          Main.hs
-  ghc-options:      -Wall -threaded
+  ghc-options:      -Wall -rtsopts -threaded
   build-depends:
     , base
     , cabal-docspec-internal
+
+test-suite cabal-docspec-tests
+  default-language: Haskell2010
+  type:             exitcode-stdio-1.0
+  hs-source-dirs:   tests
+  main-is:          tests.hs
+  ghc-options:      -Wall -rtsopts -threaded
+  build-depends:
+    , base
+    , cabal-docspec-internal
+    , containers
+    , tasty
+    , tasty-hunit

--- a/cabal-docspec/src/CabalDocspec/ExprVars.hs
+++ b/cabal-docspec/src/CabalDocspec/ExprVars.hs
@@ -1,0 +1,42 @@
+{-# LANGUAGE DeriveFunctor #-}
+module CabalDocspec.ExprVars (
+    exprVars,
+) where
+
+import Peura
+
+import qualified Language.Haskell.Lexer  as L
+import qualified Data.Set as Set
+
+exprVars :: String -> Set String
+exprVars input =
+    expr Set.empty Set.empty $ filter notWhitespace $ L.lexerPass0 input
+  where
+    expr :: Set String -> Set String -> [L.PosToken] -> Set String
+    expr  acc  bound []                                 = Set.difference acc bound
+    
+    -- collect ids
+    expr  acc  bound ((L.Varid, (_, v))         : rest) = expr (Set.insert v acc) bound rest
+    -- we skip Varsyms. Don't use them in properties :)
+
+    expr  acc  bound ((L.Reservedop, (_, "\\")) : rest) = 
+        let (vars, rest') = span notArrow rest
+        in expr acc (bound <> Set.fromList (mapMaybe isVarid vars)) rest'
+
+    -- debug
+    -- expr  acc  bound rest = error (show (take 5 rest))
+
+    -- for other tokens, we simply continue.
+    expr  acc  bound (_                         : rest) = expr acc bound rest
+
+notWhitespace :: L.PosToken -> Bool
+notWhitespace (L.Whitespace, _) = False
+notWhitespace _                 = True
+
+isVarid :: L.PosToken -> Maybe String
+isVarid (L.Varid, (_, v)) = Just v
+isVarid _                 = Nothing
+
+notArrow :: L.PosToken -> Bool
+notArrow (L.Reservedop, (_, "->")) = False
+notArrow _                         = True

--- a/cabal-docspec/src/CabalDocspec/Main.hs
+++ b/cabal-docspec/src/CabalDocspec/Main.hs
@@ -78,8 +78,8 @@ main = do
 
         res <- case optCabalPlan opts of
             CabalPlan -> do
-                builddir <- makeAbsolute (optBuilddir opts)
-                plan <- liftIO $ Plan.findAndDecodePlanJson $ Plan.InBuildDir $ toFilePath builddir
+                buildDir <- makeAbsolute (optBuildDir opts)
+                plan <- liftIO $ Plan.findAndDecodePlanJson $ Plan.InBuildDir $ toFilePath buildDir
 
                 -- checks
                 checkGhcVersion tracer ghcInfo plan
@@ -115,7 +115,7 @@ main = do
                     for (pkgUnits pkg) $ \unit ->
                         ifor (Plan.uComps unit) $ \cn ci -> do
                             testComponent tracer0 tracer (optGhci opts)
-                                ghcInfo builddir cabalCfg plan env pkg unit cn ci
+                                ghcInfo buildDir cabalCfg plan env pkg unit cn ci
 
                 -- summarize Summary's
                 return $ foldMap (foldMap (foldMap id)) res
@@ -204,7 +204,7 @@ testComponent
     -> TracerPeu r Tr
     -> (DynOpts -> DynOpts)
     -> GhcInfo
-    -> Path Absolute -- ^ builddir
+    -> Path Absolute -- ^ buildDir
     -> Cabal.Config Identity
     -> Plan.PlanJson
     -> [(String, String)]
@@ -230,8 +230,8 @@ testComponent tracer0 tracerTop dynOptsCli ghcInfo buildDir cabalCfg plan env pk
     let tracer = adjustTracer (optVerbosity dynOpts) tracer0
 
     -- find extra units
-    extraUnitIds <- findExtraPackages tracer plan (optExtraPkgs dynOpts)
-
+    extraUnitIds <- findExtraPackages tracer plan (propPkgs dynOpts ++ optExtraPkgs dynOpts)
+        
     -- find library module paths
     modulePaths <- findModules
         tracer
@@ -252,9 +252,12 @@ testComponent tracer0 tracerTop dynOptsCli ghcInfo buildDir cabalCfg plan env pk
             map toCabal (toList (Plan.ciLibDeps ci)) ++
             extraUnitIds
 
+    -- cpp include dirs
+    cppDirs <- traverse makeAbsolute (optCppIncludeDirs dynOpts)
+
     -- first phase: read modules and extract the comments
     modules <- for modulePaths $ \(modname, modpath) ->
-        phase1 tracer ghcInfo (pkgDir pkg) pkgIds bi modname modpath
+        phase1 tracer ghcInfo (pkgDir pkg) cppDirs pkgIds bi modname modpath
 
     -- extract doctests from the modules.
     let parsed :: [Module [Located DocTest]]
@@ -268,7 +271,7 @@ testComponent tracer0 tracerTop dynOptsCli ghcInfo buildDir cabalCfg plan env pk
         return $ foldMap skipModule parsed
 
 -- Skip other components
-testComponent _tracer0 _tracerTop _dynOpts _ghcInfo _builddir _cabalCfg _plan _env _pkg _unit _cn _ci =
+testComponent _tracer0 _tracerTop _dynOpts _ghcInfo _buildDir _cabalCfg _plan _env _pkg _unit _cn _ci =
     return mempty
 
 -------------------------------------------------------------------------------
@@ -325,7 +328,7 @@ testComponentNo tracer0 tracerTop dynOptsCli ghcInfo cabalCfg dbG pkg = do
     -- we don't have install plan, so we look for packages in IPI
     depends <- for (C.targetBuildDepends bi) $ \dep -> findUnit (C.depPkgName dep)
     thisUnitId <- findUnit (C.packageName (pkgGpd pkg))
-    extraUnitIds <- traverse findUnit (optExtraPkgs dynOpts)
+    extraUnitIds <- traverse findUnit (propPkgs dynOpts ++ optExtraPkgs dynOpts)
 
     let pkgIds :: [PackageIdentifier]
         pkgIds = map snd depends
@@ -341,9 +344,12 @@ testComponentNo tracer0 tracerTop dynOptsCli ghcInfo cabalCfg dbG pkg = do
         (C.hsSourceDirs bi)
         (C.exposedModules lib)
 
+    -- cpp include dirs
+    cppDirs <- traverse makeAbsolute (optCppIncludeDirs dynOpts)
+
     -- first phase: read modules and extract the comments
     modules <- for modulePaths $ \(modname, modpath) ->
-        phase1 tracer ghcInfo (pkgDir pkg) pkgIds bi modname modpath
+        phase1 tracer ghcInfo (pkgDir pkg) cppDirs pkgIds bi modname modpath
 
     -- extract doctests from the modules.
     let parsed :: [Module [Located DocTest]]
@@ -385,6 +391,11 @@ findExtraPackages tracer plan = traverse $ \pn -> do
 -------------------------------------------------------------------------------
 -- Utilities
 -------------------------------------------------------------------------------
+
+propPkgs :: DynOpts -> [PackageName]
+propPkgs dynOpts = case optProperties dynOpts of
+    SkipProperties   -> []
+    SimpleProperties -> [ mkPackageName "QuickCheck" ]
 
 manglePackageName :: C.PackageName -> String
 manglePackageName = map fixchar . prettyShow where

--- a/cabal-docspec/src/CabalDocspec/Opts.hs
+++ b/cabal-docspec/src/CabalDocspec/Opts.hs
@@ -29,6 +29,7 @@ data DynOpts = DynOpts
     , optExts           :: [String]
     , optTimeout        :: Double
     , optTimeoutMsg     :: String          -- ^ timeout response
+    , optGhciRtsopts    :: [String]
     , optSetup          :: [String]
     , optExtraPkgs      :: [PackageName]
     , optCppIncludeDirs :: [FsPath]
@@ -45,6 +46,7 @@ defaultDynOpts = DynOpts
     , optExts           = []
     , optTimeout        = 3
     , optTimeoutMsg     = "* Hangs forever *"
+    , optGhciRtsopts    = []
     , optSetup          = []
     , optExtraPkgs      = []
     , optCppIncludeDirs = []
@@ -143,6 +145,7 @@ dynOptsP = pure combine
     <*> listP extP
     <*> timeoutP
     <*> timeoutMsgP
+    <*> monoidP rtsOptsP
     <*> listP (O.strOption (O.long "setup" <> O.metavar "EXPR" <> O.help "A setup expression"))
     <*> listP extraPkgP
     <*> listP cppDirP
@@ -156,8 +159,8 @@ dynOptsP = pure combine
     monoidP :: Monoid a => O.Parser a -> O.Parser (a -> a)
     monoidP p = (\xs ys -> mconcat (ys : xs)) <$> many p
 
-    combine f1 f2 f3 f4 f5 f6 f7 f8 f9 fA fB fC (DynOpts x1 x2 x3 x4 x5 x6 x7 x8 x9 xA xB xC) =
-        DynOpts (f1 x1) (f2 x2) (f3 x3) (f4 x4) (f5 x5) (f6 x6) (f7 x7) (f8 x8) (f9 x9) (fA xA) (fB xB) (fC xC)
+    combine f1 f2 f3 f4 f5 f6 f7 f8 f9 fA fB fC fD (DynOpts x1 x2 x3 x4 x5 x6 x7 x8 x9 xA xB xC xD) =
+        DynOpts (f1 x1) (f2 x2) (f3 x3) (f4 x4) (f5 x5) (f6 x6) (f7 x7) (f8 x8) (f9 x9) (fA xA) (fB xB) (fC xC) (fD xD)
 
 lastOpt :: [a] -> a -> a
 lastOpt xs initial = foldl' (\_ x -> x) initial xs
@@ -196,6 +199,10 @@ extP = O.strOption (O.short 'X' <> O.metavar "EXT" <> O.help "Extensions")
 extraPkgP :: O.Parser PackageName
 extraPkgP = O.option (O.eitherReader C.eitherParsec) $
     O.long "extra-package" <> O.metavar "PKG" <> O.help "Extra packages to require (should exist in a plan)"
+
+rtsOptsP :: O.Parser [String]
+rtsOptsP = O.option (fmap words O.str) $
+    O.long "ghci-rtsopts" <> O.metavar "OPTS" <> O.help "RTS options for GHCi process"
 
 propVariablesP :: O.Parser (Set String)
 propVariablesP = O.option (fmap (Set.fromList . words) O.str) $

--- a/cabal-docspec/src/CabalDocspec/Opts.hs
+++ b/cabal-docspec/src/CabalDocspec/Opts.hs
@@ -2,55 +2,66 @@ module CabalDocspec.Opts where
 
 import Peura
 
-import qualified Distribution.Parsec          as C
-import qualified Distribution.Types.BuildInfo as C
+import qualified Data.Set                        as Set
 import qualified Distribution.Compat.CharParsing as P
-import qualified Options.Applicative          as O
+import qualified Distribution.Parsec             as C
+import qualified Distribution.Types.BuildInfo    as C
+import qualified Options.Applicative             as O
 
 import CabalDocspec.Trace
 import CabalDocspec.Warning
 
 data Opts = Opts
-    { optCabalPlan  :: CabalPlan
-    , optGhci       :: DynOpts -> DynOpts
-    , optCompiler   :: FilePath
-    , optBuilddir   :: FsPath
-    , optTracer     :: TracerOptions W -> TracerOptions W
-    , optTargets    :: [String]
+    { optCabalPlan :: CabalPlan
+    , optGhci      :: DynOpts -> DynOpts
+    , optCompiler  :: FilePath
+    , optBuildDir  :: FsPath
+    , optTracer    :: TracerOptions W -> TracerOptions W
+    , optTargets   :: [String]
     }
 
 
 -- | Options which can change per component.
 data DynOpts = DynOpts
-    { optPhase      :: Phase
-    , optPreserveIt :: PreserveIt
-    , optStripComs  :: StripComments
-    , optExts       :: [String]
-    , optTimeout    :: Double
-    , optTimeoutMsg :: String          -- ^ timeout response
-    , optSetup      :: [String]
-    , optExtraPkgs  :: [PackageName]
-    , optVerbosity  :: Verbosity
+    { optPhase          :: Phase
+    , optPreserveIt     :: PreserveIt
+    , optStripComs      :: StripComments
+    , optExts           :: [String]
+    , optTimeout        :: Double
+    , optTimeoutMsg     :: String          -- ^ timeout response
+    , optSetup          :: [String]
+    , optExtraPkgs      :: [PackageName]
+    , optCppIncludeDirs :: [FsPath]
+    , optProperties     :: Properties
+    , optPropVariables  :: Set String
+    , optVerbosity      :: Verbosity
     }
-  deriving Show
 
 defaultDynOpts :: DynOpts
 defaultDynOpts = DynOpts
-    { optPhase      = Phase2
-    , optPreserveIt = DontPreserveIt
-    , optStripComs  = DontStripComments
-    , optExts       = []
-    , optTimeout    = 3
-    , optTimeoutMsg = "* Hangs forever *"
-    , optSetup      = []
-    , optExtraPkgs  = []
-    , optVerbosity  = Verbosity 0
+    { optPhase          = Phase2
+    , optPreserveIt     = DontPreserveIt
+    , optStripComs      = DontStripComments
+    , optExts           = []
+    , optTimeout        = 3
+    , optTimeoutMsg     = "* Hangs forever *"
+    , optSetup          = []
+    , optExtraPkgs      = []
+    , optCppIncludeDirs = []
+    , optProperties     = SkipProperties
+    , optPropVariables  = mempty
+    , optVerbosity      = Verbosity 0
     }
 
 newtype Verbosity  = Verbosity Int deriving (Eq, Ord, Show)
 data PreserveIt    = PreserveIt | DontPreserveIt deriving (Eq, Show)
 data CabalPlan     = CabalPlan | NoCabalPlan deriving (Eq, Show)
 data StripComments = StripComments | DontStripComments deriving (Eq, Show)
+
+data Properties
+    = SkipProperties
+    | SimpleProperties
+  deriving (Eq, Ord, Show)
 
 data Phase
     = Manual
@@ -84,6 +95,16 @@ dynOptsFromBuildInfo tracer bi = do
                     putWarning tracer WInvalidField $ name ++ ": optparse-applicatice completion invoked"
                     return id
 
+    parse name@"x-docspec-property-variables" contents =
+        case C.explicitEitherParsec (many (C.parsecToken <* P.spaces)) contents of
+            Left err -> do
+                putWarning tracer WInvalidField $ name ++ ": " ++ err
+                return id
+            
+            Right strs -> return $ \dynOpts -> dynOpts
+                { optPropVariables = Set.fromList strs <> optPropVariables dynOpts
+                }
+
     parse name@"x-docspec-extra-packages" contents =
         case C.explicitEitherParsec (many (C.parsec <* P.spaces)) contents of
             Left err -> do
@@ -110,8 +131,9 @@ optsP = pure Opts
     <*> O.option fspath (O.long "builddir" <> O.value (fromFilePath "dist-newstyle") <> O.metavar "BUILDDIR")
     <*> tracerOptionsParser
     <*> many (O.strArgument $ O.metavar "TARGET")
-  where
-    fspath = O.eitherReader $ return . fromFilePath
+
+fspath :: O.ReadM FsPath
+fspath = O.eitherReader $ return . fromFilePath
 
 dynOptsP :: O.Parser (DynOpts -> DynOpts)
 dynOptsP = pure combine
@@ -123,11 +145,19 @@ dynOptsP = pure combine
     <*> timeoutMsgP
     <*> listP (O.strOption (O.long "setup" <> O.metavar "EXPR" <> O.help "A setup expression"))
     <*> listP extraPkgP
+    <*> listP cppDirP
+    <*> propertiesP
+    <*> monoidP propVariablesP
     <*> verbosityP
   where
+    listP :: O.Parser a -> O.Parser ([a] -> [a])
     listP p = flip (++) <$> many p
-    combine f1 f2 f3 f4 f5 f6 f7 f8 f9 (DynOpts x1 x2 x3 x4 x5 x6 x7 x8 x9) =
-        DynOpts (f1 x1) (f2 x2) (f3 x3) (f4 x4) (f5 x5) (f6 x6) (f7 x7) (f8 x8) (f9 x9)
+
+    monoidP :: Monoid a => O.Parser a -> O.Parser (a -> a)
+    monoidP p = (\xs ys -> mconcat (ys : xs)) <$> many p
+
+    combine f1 f2 f3 f4 f5 f6 f7 f8 f9 fA fB fC (DynOpts x1 x2 x3 x4 x5 x6 x7 x8 x9 xA xB xC) =
+        DynOpts (f1 x1) (f2 x2) (f3 x3) (f4 x4) (f5 x5) (f6 x6) (f7 x7) (f8 x8) (f9 x9) (fA xA) (fB xB) (fC xC)
 
 lastOpt :: [a] -> a -> a
 lastOpt xs initial = foldl' (\_ x -> x) initial xs
@@ -155,12 +185,21 @@ stripComsP = lastOpt <$> many (on <|> off) where
     on  = O.flag'     StripComments $ O.long    "strip-comments" <> O.help "Strip comments in examples"
     off = O.flag' DontStripComments $ O.long "no-strip-comments" <> O.help "Don't strip comments in examples"
 
+propertiesP :: O.Parser (Properties -> Properties)
+propertiesP = lastOpt <$> many (skip <|> simple) where
+    skip   = O.flag' SkipProperties   $ O.long "skip-properties"   <> O.help "Skip properties (default)"
+    simple = O.flag' SimpleProperties $ O.long "simple-properties" <> O.help "SImple properties: evaluate using quickCheck (prop expr)"
+
 extP :: O.Parser String
 extP = O.strOption (O.short 'X' <> O.metavar "EXT" <> O.help "Extensions")
 
 extraPkgP :: O.Parser PackageName
 extraPkgP = O.option (O.eitherReader C.eitherParsec) $
     O.long "extra-package" <> O.metavar "PKG" <> O.help "Extra packages to require (should exist in a plan)"
+
+propVariablesP :: O.Parser (Set String)
+propVariablesP = O.option (fmap (Set.fromList . words) O.str) $
+    O.long "property-variables" <> O.metavar "VARS" <> O.help "Variables to automatically quantify over in properties"
 
 phaseP :: O.Parser (Phase -> Phase)
 phaseP = lastOpt <$> many (phase1P <|> phase2P <|> manualP) where
@@ -178,3 +217,6 @@ verbosityP = accum <$> many (plusP <|> minusP)
 
     minusP :: O.Parser Int
     minusP = O.flag' (negate 1) (O.short 'q' <> O.long "quiet" <> O.help "Decrease verbosity level")
+
+cppDirP :: O.Parser FsPath
+cppDirP = O.option fspath (O.short 'I' <> O.help "Add ⟨dir⟩ to the directory search list for #include files")

--- a/cabal-docspec/src/CabalDocspec/Phase1.hs
+++ b/cabal-docspec/src/CabalDocspec/Phase1.hs
@@ -23,12 +23,13 @@ phase1
     :: TracerPeu r Tr
     -> GhcInfo
     -> Path Absolute       -- ^ package directory
+    -> [Path Absolute]     -- ^ additional include directories
     -> [PackageIdentifier] -- ^ dependencies
     -> C.BuildInfo
     -> C.ModuleName
     -> Path Absolute
     -> Peu r (Module (Located String))
-phase1 tracer ghcInfo pkgDir pkgIds bi modname modpath = do
+phase1 tracer ghcInfo pkgDir cppDirs pkgIds bi modname modpath = do
     traceApp tracer $ TracePhase1 modname modpath
 
     contents <- fromUTF8BS <$> readByteString modpath
@@ -52,7 +53,7 @@ phase1 tracer ghcInfo pkgDir pkgIds bi modname modpath = do
         -- so may break
         [ pkgDir </> fromUnrootedFilePath dir
         | dir <- C.includeDirs bi
-        ]
+        ] ++ cppDirs
 
     cppDefines :: [(String, String)]
     cppDefines =

--- a/cabal-docspec/src/CabalDocspec/Phase2.hs
+++ b/cabal-docspec/src/CabalDocspec/Phase2.hs
@@ -7,15 +7,18 @@ module CabalDocspec.Phase2 (
 import Peura
 
 import Control.Monad      (foldM)
+import Data.List          (isInfixOf)
 import System.Environment (getEnvironment)
 
 import qualified Cabal.Config           as Cabal
 import qualified Data.Map               as Map
+import qualified Data.Set               as Set
 import qualified Language.Haskell.Lexer as L
 
 import CabalDocspec.Doctest.Example
 import CabalDocspec.Doctest.Extract
 import CabalDocspec.Doctest.Parse
+import CabalDocspec.ExprVars
 import CabalDocspec.GHCi
 import CabalDocspec.Located
 import CabalDocspec.Opts
@@ -141,9 +144,33 @@ phase2 tracer dynOpts unitIds ghcInfo mbuildDir cabalCfg cwd extraEnv parsed = d
 
             let runExample :: Summary -> Located DocTest -> Peu r (Either Summary Summary)
                 runExample acc (L pos doctest) = case doctest of
-                    Property expr -> do
-                        putError tracer $ "properties not implemented, skipping " ++ expr
-                        return (Left (acc <> mempty { sProperties = ssSkip }))
+                    Property expr -> case optProperties dynOpts of
+                        SkipProperties -> do
+                            putWarning tracer WSkippedProperty $ prettyPos pos ++ " skipping " ++ expr
+                            return (Left (acc <> mempty { sProperties = ssSkip }))
+
+                        SimpleProperties -> do
+                            let vars = exprVars expr `Set.intersection` optPropVariables dynOpts
+
+                            let lambdaPrefix :: String
+                                lambdaPrefix
+                                    | null vars = ""
+                                    | otherwise = unwords $ "\\" : Set.toList vars ++ ["-> "]
+
+                            -- imports
+                            _ <- eval tracer ghci False fastTimeout timeoutMsg
+                                "import Test.QuickCheck (quickCheck)"
+
+                            -- evaluate property
+                            result <- eval tracer ghci False timeout timeoutMsg $
+                                "quickCheck (" ++ lambdaPrefix ++ expr ++ ")"
+
+                            if "OK, passed" `isInfixOf` result
+                            then return (Left (acc <> mempty { sProperties = ssSuccess }))
+                            else do
+                                putError tracer expr
+                                putError tracer $ prettyPos pos ++ "\n" ++ result
+                                return (Right (acc <> mempty { sProperties = ssError }))
 
                     Example expr expected -> do
                         result <- eval tracer ghci preserveIt timeout timeoutMsg expr

--- a/cabal-docspec/src/CabalDocspec/Phase2.hs
+++ b/cabal-docspec/src/CabalDocspec/Phase2.hs
@@ -56,6 +56,11 @@ phase2 tracer dynOpts unitIds ghcInfo mbuildDir cabalCfg cwd extraEnv parsed = d
 
     let GhcFlags {..} = getGhcFlags ghcInfo
 
+    let rtsArgs :: [String]
+        rtsArgs
+            | null (optGhciRtsopts dynOpts) = []
+            | otherwise = ["+RTS"] ++ optGhciRtsopts dynOpts ++ ["-RTS"]
+
     let ghciArgs :: [String]
         ghciArgs =
             [ "-i" -- so we don't explode on hs-source-dirs: . packages
@@ -76,7 +81,7 @@ phase2 tracer dynOpts unitIds ghcInfo mbuildDir cabalCfg cwd extraEnv parsed = d
             ] ++
             [ ghcFlagPackageId ++ "=" ++ u
             | u <- map prettyShow unitIds
-            ]
+            ] ++ rtsArgs
 
     currEnv <- liftIO getEnvironment
     let env = Map.toList $ Map.fromList $ extraEnv ++ currEnv

--- a/cabal-docspec/src/CabalDocspec/Warning.hs
+++ b/cabal-docspec/src/CabalDocspec/Warning.hs
@@ -10,6 +10,7 @@ data W
     | WInvalidField
     | WCpphs
     | WErrorInSetup
+    | WSkippedProperty
   deriving (Eq, Ord, Enum, Bounded)
 
 instance Universe W where universe = [minBound .. maxBound]
@@ -23,3 +24,4 @@ instance Warning W where
     warningToFlag WInvalidField        = "invalid-field"
     warningToFlag WCpphs               = "cpphs"
     warningToFlag WErrorInSetup        = "error-in-setup"
+    warningToFlag WSkippedProperty     = "skipped-property"

--- a/cabal-docspec/tests/tests.hs
+++ b/cabal-docspec/tests/tests.hs
@@ -1,0 +1,22 @@
+module Main (main) where
+
+import Test.Tasty (defaultMain, testGroup, TestTree)
+import Test.Tasty.HUnit (testCase, (@=?))
+
+import qualified Data.Set as Set
+
+import CabalDocspec.ExprVars
+
+main :: IO ()
+main = defaultMain $ testGroup "cabal-docspec"
+    [ exprVarsTests
+    ]
+
+exprVarsTests :: TestTree
+exprVarsTests = testGroup "ExprVars"
+    [ ex "x + y"                                 ["x","y"]
+    , ex "forAll xs $ \\x -> x == x"             ["forAll", "xs"]
+    , ex "\\(xs :: [Int]) -> reverse xs === xs"  ["reverse"]
+    ]
+  where
+    ex expr vars = testCase expr $ Set.fromList vars @=? exprVars expr

